migration(issue-views): Add new `projects`, `environments`, `time_filter` columns to `GroupSearchView`

MichaelSun48 opened the pull request:
This PR adds new columns to the GroupSearchView table to represent the new filters we'd like to add to Issue Views: `projects`, `is_my_projects`, `environments`, and `time_filters`. This will support a new feature in Issue Views where users can save projects, environment, and time filters to their views, rather than it being a global filter.

Commit:
Add new columns to represent page filters

Commit:
Use correct function parameter for default

Commit:
Use DefaultFieldsModel

Commit:
Set default for is_my_projects to true

MichaelSun48 commented on src/sentry/models/groupsearchview.py:
```diff
@@ -26,6 +56,13 @@ class GroupSearchView(DefaultFieldsModelExisting):
     )
     position = models.PositiveSmallIntegerField()
 
+    projects = models.ManyToManyField("sentry.Project", through="sentry.GroupSearchViewProject")
+    is_my_projects = models.BooleanField(db_default=True)
+    environments = models.ManyToManyField(
+        "sentry.Environment", through="sentry.GroupSearchViewEnvironment"
+    )
```
Going with ManyToMany fields instead of ArrayFields here - I was a little too adamant about trying to make these fields non-nullable to conceptually agree with the requirement that views _must_ have page filters, but it really doesn't matter that much since we can just equate a null field to equal "All", which is the chosen default anyways. This also means we won't have to worry about any orphaned data in Arrays.

Commit:
use db_defualt instead of default

Commit:
Use plain dict for json instead of function

JoshFerge reviewed (approved):
makes sense to me.

MichaelSun48 commented:
The backend tests are failing on `tests/sentry/backup/test_exhaustive.py` because it's not finding the intermediate tables created by the M2M fields (`GroupSearchViewProjects` and `GroupSearchViewEnvironments`), but if I'm undestanding correctly based on [this line](https://github.com/getsentry/sentry/blob/f9c57063871d8d1885ab9cb48ef99366611c89cd/tests/sentry/backup/__init__.py#L84), these fields should not be included in that test. Can someone confirm if that's the case?

wedamija commented:
> The backend tests are failing on `tests/sentry/backup/test_exhaustive.py` because it's not finding the intermediate tables created by the M2M fields (`GroupSearchViewProjects` and `GroupSearchViewEnvironments`), but if I'm undestanding correctly based on [this line](https://github.com/getsentry/sentry/blob/f9c57063871d8d1885ab9cb48ef99366611c89cd/tests/sentry/backup/__init__.py#L84), these fields should not be included in that test. Can someone confirm if that's the case?

I think that it's skipping the M2M field itself, but those models still exist individually, and so should have to be included in the test.

wedamija commented on src/sentry/models/groupsearchview.py:
```diff
@@ -26,6 +52,13 @@ class GroupSearchView(DefaultFieldsModelExisting):
     position = models.PositiveSmallIntegerField()
 
+    projects = models.ManyToManyField("sentry.Project", through="sentry.GroupSearchViewProject")
+    is_my_projects = models.BooleanField(db_default=True)
```
So if this is True, then we don't need to look at the list of projects? And if `projects` has no results, then we assume all?

MichaelSun48 replied:
That's correct.

MichaelSun48 replied:
That reminds me, this is an important thing to have commented directly in the model, adding that now

wedamija replied:
One edge case here: This view has one project, it is deleted, and now this view switches from being filtered to that project to all projects.

Would it be more flexible if this was `mode` instead, and we have options:
 - my_projects
 - all_projects
 - specified_projects (don't necessarily use this name, but this idea)

MichaelSun48 replied:
Good point - but doesn't this still run into the same issue? We'd need some trigger that checks changes the `mode` if the only project is deleted. 

Could we just map no results to "My Projects" instead?

wedamija replied:
That's a good point, we should just have a post-delete signal on the m2m table that deletes the related row if there are no rows left, and leave it as it is. Then there's no problem

wedamija commented on src/sentry/models/groupsearchview.py:
```diff
@@ -26,6 +52,13 @@ class GroupSearchView(DefaultFieldsModelExisting):
     position = models.PositiveSmallIntegerField()
 
+    projects = models.ManyToManyField("sentry.Project", through="sentry.GroupSearchViewProject")
+    is_my_projects = models.BooleanField(db_default=True)
```
Just checking - since we don't have projects today, we're currently assuming that it's always `my_projects`?

MichaelSun48 replied:
That's right - that's the intended default. There'll also be some logic on the GET request that accesses this model to checks if the organization has the `global-views` feature and corrects the column if they don't

Commit:
Add instances of bridge tables to backups

Commit:
Fix type in exhaustive fixture

Commit:
Change to is_all_projects instead of is_my_projects

Commit:
Fix backups typing again

Commit:
Add comparators, rebase on master

Commit:
Fix date comparators AGAIN

Commit:
Set is_all_projects default to be false

Commit:
Turn environemnts into an array field

MichaelSun48 commented:
Hey @wedamija, mind if I get one last sanity check? Made some non trivial changes since the last review, detailed here: 
1. `is_my_projects` has been turned into `is_all_projects`, and is now set to `False` by default. No projects in `GroupSearchViewProjects` now equates to "My Projects", instead of "All Projects". This is so when the only project gets deleted, the view will inherently fallback to the default, "My Projects" 
2. `environments` has been turned into an `ArrayField(CharField(...), ...)`. This is because the frontend doesn't actually use the Environments model, and instead just gets environments off of the projects endpoint (see screenshot). This isn't ideal imo, but it does save a lot of frontend refactoring, and fwiw, is what the `Dashboard` model does.

![Screenshot 2025-01-15 at 11 18 42 AM](https://github.com/user-attachments/assets/fd1ad758-96b7-4ebb-ba2a-9fcfdc4810ce)

wedamija commented on src/sentry/backup/comparators.py:
```diff
@@ -806,6 +806,10 @@ def get_default_comparators() -> dict[str, list[JSONScrubbingComparator]]:
                 DateUpdatedComparator("date_added", "date_updated"),
             ],
             "sentry.groupsearchview": [DateUpdatedComparator("date_updated")],
+            "sentry.groupsearchviewproject": [
+                DateUpdatedComparator("date_updated"),
+                DateUpdatedComparator("date_added"),
+            ],
```
Hmm, this is annoying that we still need to do it sometimes... I'll try look into fixing it

wedamija commented on src/sentry/models/groupsearchview.py:
```diff
@@ -26,6 +41,16 @@ class GroupSearchView(DefaultFieldsModelExisting):
+    is_all_projects = models.BooleanField(db_default=False)
+    # Environments = [] maps to "All Environments"
+    environments = ArrayField(
+        models.CharField(max_length=ENVIRONMENT_NAME_MAX_LENGTH, null=False, db_default=[])
```
The `db_default` should be on the `ArrayField`

wedamija commented:
> Hey @wedamija, mind if I get one last sanity check? Made some non trivial changes since the last review, detailed here:
> 
> 1. `is_my_projects` has been turned into `is_all_projects`, and is now set to `False` by default. No projects in `GroupSearchViewProjects` now equates to "My Projects", instead of "All Projects". This is so when the only project gets deleted, the view will inherently fallback to the default, "My Projects"
> 2. `environments` has been turned into an `ArrayField(CharField(...), ...)`. This is because the frontend doesn't actually use the Environments model, and instead just gets environments off of the projects endpoint (see screenshot). This isn't ideal imo, but it does save a lot of frontend refactoring, and fwiw, is what the `Dashboard` model does.
> 
> ![Screenshot 2025-01-15 at 11 18 42 AM](https://private-user-images.githubusercontent.com/55160142/403555539-fd1ad758-96b7-4ebb-ba2a-9fcfdc4810ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY5NzAzMjUsIm5iZiI6MTczNjk3MDAyNSwicGF0aCI6Ii81NTE2MDE0Mi80MDM1NTU1MzktZmQxYWQ3NTgtOTZiNy00ZWJiLWJhMmEtOWZjZmRjNDgxMGNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE1VDE5NDAyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmOGQ0MTQ5YWM0NWM5MzQxNzdmZGVkMjVhZGMxZjlmOGYxZGRiNzlhNWM0Njc4Y2I2NTlmZjExOWVhYWY1ZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.M_mX25cV-NF2tVxgtbSkhMQKoDLIdD4wbdN2oIei96k)

I don't feel too strongly here, but probably we shouldn't design the backend tables based on how the frontend uses it. As an alternative, could you use the m2m table, but in the serializer return the list of env names? Either way, not blocking on this, just a suggestion

MichaelSun48 commented:
> I don't feel too strongly here, but probably we shouldn't design the backend tables based on how the frontend uses it. As an alternative, could you use the m2m table, but in the serializer return the list of env names? Either way, not blocking on this, just a suggestion

@wedamija 
That was my first thought too, but an issue would arise once we tried to send the environments back to be saved. The views would get the environments to save from the [The environment dropdown](https://github.com/getsentry/sentry/blob/master/static/app/components/organizations/environmentPageFilter/index.tsx), but that component doesn't keep track of the environment primary keys, only the environment names.

Commit:
Move db_default to array

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

Commit:
Fix field default again again

MichaelSun48 commented:
Reminder to self: make global consts for default values, they will be reused.

Commit:
Create global const for default time filter
